text/0074-global-init.md

```diff
@@ -0,0 +1,176 @@
+provider (factory) instances.  This OTEP proposes a specification for how to
+initialize the global provider (factory) instances.
+
+In some programming environments, it is possible for libraries of code
+to auto-initialize static variables, allowing them to begin operation
+concurrently with the main program, while initializing static program
+state.  This presents a set of opposing requirements: (1) the API
+supports a configurable SDK; (2) third party libraries can use
+OpenTelemetry before the SDK is configured.
+
+The current specification discusses a global provider (factory) for
+named Tracers and Meters as well as a global Propagators instance for
+injectors and extractors, but does not discuss how these are
+initialized or whether their values can be modified during the process
+lifetime.
+
+Global variables face significant opposition from some developers,
+which forces the question: "why support globals at all?".  In
+languages with automatic dependency injection support, then
+conceivably we do not need global variables.  In languages without
+automatic dependency injection, without globals we could not have the
+"zero touch" instrumentation [given as a requirement for the
+project](https://github.com/open-telemetry/oteps/blob/master/text/0001-telemetry-without-manual-instrumentation.md).
+If a third-party library is to be instrumented and integrated without
+modification into an application, either dependency injection or
+global variables are the solution.
+
+Global variables **are a dangerous programming pattern**, but they
+also enable easy integration in languages without automatic dependency
+injection.  To address this risk, this proposal specifies strict
+limits on their initialization.  We propose at-most-once
+initialization for the three global variables in OpenTelemetry.
+Specifically, this specification says that the global Tracer provider,
+the global Meter provider, and the global Propagators instance can
+only be initialized once per process lifetime (except possibly in
+test-only scenarios).
+
+This OTEP also specifies the behavior of the global instances when
+they are used before the SDK is configured and installed, in case this
+cannot be performed by automatic dependency injection.
+
+## Explanation
+
+There are two acceptable ways to provide default instances in
+OpenTelemetry: (1) through dependency injection, (2) through global
+variables initialized at most once.  The feasibility of each approach
+varies by language.  The implementation MUST select one of the
+following strategies.
+
+### Service provider mechanism
+
+Where the language provides a commonly accepted way to inject SDK
+components, it should be preferred.  The Java SPI (Service Provider
+Interface) supports loading and configuring the SDK before it is first
+used.  This kind of support is the preferred choice in languages with
+common support for automatic dependency injection.
+
+In this case, there is no use-before-configuration question to
```

Comment:
I am not sure this use case is so much different from the next section. Even though Java often uses SPI, it's not a requirement that I would expect to be baked-in into OpenTelemetry Java API/SDK, i.e. if someone wants to instantiate OTel providers explicitly, that should still be an option, regardless of additional support for SPI.

Reply:
This is really all I'm asking for. Let's make sure we don't force SPI for people who don't want to use SPI, and want to manually configure their applications.

Reply:
(note: It's currently baked into the Java SDK, and it's going to take a bit of elbow grease to unwind it at this point)

Reply:
(I'd like to see a prototype of Java's `io.opentelemetry.OpenTelemetry` supporting both scenarios - else, we will have to stick, at least for such global object, to **either** SPI **or** manual initialization. And observe this is in the API side, which will definitely impact the SDK).

Reply:
Support for both mechanisms sounds difficult to me, but possible. I would suggest that the dependency injection mechanism be allowed to provide a default instance, in which case it is initialized before use and no forwarding mechanism enters play. If DI does not provide an implementation, then the global "forwarding" implementation is used (with the same requirements as stated in this document) and the user can explicitly initialize the SDK in its `Main()`. 

It's been my intention to restrict the use of globals as much as possible, which is where the "only once" requirement comes from.  If you permit any caller to reset the global SDK at any time, it seems to invite trouble.

FWIW, I did a a Golang prototype that used the `plugin` package to "inject" an implementation before it is used. This was met with frowns in the Go community, so I abandoned it. See https://github.com/jmacd/opentelemetry-go/pull/1